src/message_queue.rs

```diff
@@ -65,4 +66,34 @@ impl MessageQueue {
         };
         Ok(queue)
     }
+
+    /// Retrieve (and remove) all messages of a given type from a given sender from the MessageQueue.
+    pub(crate) fn retrieve(
+        &mut self,
+        message_type: MessageType,
+        identifier: Identifier,
+        sender: ParticipantIdentifier,
+    ) -> Result<Vec<Message>> {
+        let message_index = MessageIndex {
+            message_type,
+            identifier,
+        };
+        let key = serialize!(&message_index)?;
+        let queue = match self.0.get_mut(&key) {
+            Some(a) => a,
+            None => return Ok(vec![]),
+        };
+        let mut out_messages = Vec::new();
+        let mut indexes_to_remove = Vec::new();
+        for (i, message) in queue.iter_mut().enumerate() {
+            if message.from() == sender {
+                indexes_to_remove.push(i);
+                out_messages.push(message.clone());
+            }
+        }
+        for i in indexes_to_remove.iter().rev() {
+            queue.remove(*i);
+        }
```

Comment:
how about something like
while i != queue.len() {
    if some_predicate(&mut vec[i]) {
        let val = vec.remove(i);
        // your code here
    } else {
        i += 1;
    }
}

Reply:
I didn't quite realize when reading this that we're basically trying to split the parent `queue` into two pieces -- one that we retain as `queue` and one that we output as `out_messages`. My original suggestion was something like:

```rust
let out_message: Vec<_> = queue
    .iter()
    .filter(|msg| msg.from() == sender)
    .collect()
```

I threw it into the rust playground and it doesn't change parent queue: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0dffb17a9877c18605948835ce71fee4

However, we _do_ want it to change the parent queue; I didn't account for that. Maybe something like,
```rust
for msg in &out_messages {
    queue.remove_entry(msg);
}
```

This doesn't fix the two-pass inefficiency, but it does look Rustier to me.

Reply:
Looking at this a little more, I don't think it quite maps to the correct data structure.

Reply:
Oh, actually. The parent thing you're trying to change is the value in a hashmap, but it's type is `Vec`. So you can use `partition, which is a stable iter method to turn the `queue` into two vectors, one of which has the messages you want to keep, and one the messages you want to remove. Try:

```rust
// remove old queue
let queue = match self.0.remove(&key) {
    Some(a) => a,
    None => return Ok(vec![]),
};

// separate messages we want to retrieve
let (out, new_queue): (Vec<_>, Vec<_>) = queue
    .into_iter() 
    .partition(|msg| msg.from() == sender);

// re-add updated queue
self.0.insert(key, new_queue);

Ok(out)
```

As an aside, it would be nice to have a unit test for these methods so we could easily tell if they work :)

Reply:
This seems to work quite nicely, thanks!

Reply:
good to know, thanks!